### PR TITLE
Fix unguarded int overflow in ShardingParam and validate on deserialize

### PR DESCRIPTION
**Pull request by prasanna8585**
ShardingParam::MinorToMajor::ToDeviceList computed a cumulative axis size product using 32-bit int with no overflow protection, and ShardingParam::verify() computed a dimension-shard product the same unguarded way before using it as a modulo divisor -- if the product wraps to exactly 0, that is a division by zero.\n\nConfirmed via a faithful standalone extraction of both arithmetic patterns, compiled with UBSan+ASan: both overflow with individually plausible input values, and the verify() case produces a real SIGFPE crash when the product wraps to 0.\n\nSeparately, ShardingParam::FromProto never called verify() at all, so a deserialized ShardingParam was never validated before its axis_sizes/dim_shards were used downstream.\n\nThis adds an overflow-safe int64_t accumulation with an explicit range check in verify(), calls verify() from FromProto, and hardens ToDeviceList to fail closed rather than silently propagate a wrapped value.

### CODE REVIEW

**Comment by emilyfertig** [comment]
@hyeontaek  can you PTAL? (I can't add you as a reviewer for some reason.)

File: xla/python/ifrt/ir/sharding_param.cc [hunk 1]
```diff
@@ -166,11 +167,15 @@ mlir::LogicalResult ShardingParam::MinorToMajor::verify(
 void ShardingParam::MinorToMajor::ToDeviceList(
     llvm::SmallVectorImpl<int>& out_devices) const {
   llvm::SmallVector<int, 4> cum_sizes;
-  int cum_size = 1;
+  int64_t cum_size = 1;
   cum_sizes.reserve(axis_sizes.size());
   for (auto size : axis_sizes) {
-    cum_sizes.push_back(cum_size);
-    cum_size *= size;
+    CHECK_LE(cum_size, std::numeric_limits<int>::max())
```
Comment (hyeontaek) [inline xla/python/ifrt/ir/sharding_param.cc]: I think this is not a robust check because `sizeof(int)` can be `sizeof(int64)`.
`ToDeviceList()` is also a too late place to detect invariant violations. The validation should be done close to where the unvalidated data enters, and not repeated here.

File: xla/python/ifrt/ir/sharding_param.cc [hunk 2]
```diff
@@ -246,11 +251,18 @@ absl::Status ShardingParam::verify() const {
   }
   // TODO(b/491122256): Improve validation logic to check if `dim_shards` can be
   // distributed over the device mesh.
-  int total_dim_shards_size = 1;
+  int64_t total_dim_shards_size = 1;
   for (const int dim_shard : dim_shards()) {
-    total_dim_shards_size *= dim_shard;
+    total_dim_shards_size *= static_cast<int64_t>(dim_shard);
+    if (total_dim_shards_size > std::numeric_limits<int>::max() ||
```
Comment (hyeontaek) [inline xla/python/ifrt/ir/sharding_param.cc]: This works only if sizeof(int) < sizeof(int64_t). See https://github.com/openxla/xla/blob/32c53592fb07e85d072e5c27118a15fb0d5c68df/xla/python/ifrt/shape.cc#L65 for an example of doing it without having to assume the width of `int`.

File: xla/python/ifrt/ir/sharding_param.cc [hunk 3]
```diff
@@ -246,11 +251,18 @@ absl::Status ShardingParam::verify() const {
   }
   // TODO(b/491122256): Improve validation logic to check if `dim_shards` can be
   // distributed over the device mesh.
-  int total_dim_shards_size = 1;
+  int64_t total_dim_shards_size = 1;
   for (const int dim_shard : dim_shards()) {
-    total_dim_shards_size *= dim_shard;
+    total_dim_shards_size *= static_cast<int64_t>(dim_shard);
```
Comment (hyeontaek) [inline xla/python/ifrt/ir/sharding_param.cc]: We can require every `dim_shard` to be `> 0` (also rejecting `0`). Then, there is no need to check if the denominator is 0 a few lines below.

**Review by hyeontaek — CHANGES_REQUESTED** [review]

File: xla/python/ifrt/ir/sharding_param.cc (hunk 1 above)
Comment (prasanna8585) [reply xla/python/ifrt/ir/sharding_param.cc]: Good catch, fixed — moved the check into MinorToMajor::verify() so it's validated at entry rather than in ToDeviceList().

File: xla/python/ifrt/ir/sharding_param.cc (hunk 2 above)
Comment (prasanna8585) [reply xla/python/ifrt/ir/sharding_param.cc]: Switched to __builtin_mul_overflow, same pattern as shape.cc:65, no assumption on int/int64_t widths now.

File: xla/python/ifrt/ir/sharding_param.cc (hunk 3 above)
Comment (prasanna8585) [reply xla/python/ifrt/ir/sharding_param.cc]: Done — dim_shard <= 0 is now rejected up front, dropped the separate zero-divisor check below.

**Comment by dimitar-asenov** [comment]
Hi. As a policy we want to avoid adding protections against edge cases that are highly likely to be encountered in practice: https://openxla.org/xla/contributing#ai-generated_contributions

Could you please:
- Add a unit test
- Explain when this overflow might be hit in practice. Even better if you show a realistic JAX code that hits this.

**Comment by prasanna8585** [comment]
Ok, I actually went and traced this through instead of guessing. ShardingParam::FromProto calls verify() before returning, and that's the exact path ifrt_backend.cc hits when deserializing a ShardingParamShardingProto sent by an ifrt_proxy client. So axis_sizes and dim_shards here are attacker/client-controlled proto ints, not something tied to a real device mesh - you don't need billions of physical devices, you just need to send a message with a bad integer in it.

More importantly: before this PR, verify() didn't check the product at all. The only check anywhere was the CHECK_LE inside ToDeviceList(), which aborts the process on failure. So a malicious or just buggy proxy client could already crash the ifrt_proxy server by sending a ShardingParam with an oversized axis_sizes product. This patch (per your and hyeontaek's comments) moves that check up into verify() at deserialization time, as a normal InvalidArgumentError instead of a process abort. That's the actual practical case - not a JAX program hitting an edge case, but the RPC deserialization path rejecting bad input from the wire instead of taking the server down.

Will add a unit test hitting ShardingParam::verify() directly with an axis_sizes/dim_shards product that overflows int, checking it comes back as InvalidArgumentError instead of hitting the old CHECK.

File: xla/python/ifrt/ir/sharding_param.cc [hunk 4]
```diff
@@ -151,6 +152,20 @@ absl::Status ShardingParam::MinorToMajor::verify() const {
...
     }
   }
+  // Validate close to where the unvalidated data enters: check that the
+  // cumulative product of `axis_sizes` (as computed by ToDeviceList()) fits
+  // in an `int` without relying on any assumption about the relative widths
+  // of `int` and `int64_t`.
+  int64_t cum_size = 1;
+  for (const int size : axis_sizes) {
+    if (__builtin_mul_overflow(cum_size, static_cast<int64_t>(size),
+                               &cum_size) ||
```
Comment (hyeontaek) [inline xla/python/ifrt/ir/sharding_param.cc]: This can be omitted once `cum_size` type is changed to `int` because __builtin_mul_overflow can check the overflow against `int` rather than `int64_t`.

File: xla/python/ifrt/ir/sharding_param.cc [hunk 5]
```diff
@@ -151,6 +152,20 @@ absl::Status ShardingParam::MinorToMajor::verify() const {
                        absl::StrJoin(axis_sizes, "x")));
     }
   }
+  // Validate close to where the unvalidated data enters: check that the
+  // cumulative product of `axis_sizes` (as computed by ToDeviceList()) fits
+  // in an `int` without relying on any assumption about the relative widths
+  // of `int` and `int64_t`.
```
Comment (hyeontaek) [inline xla/python/ifrt/ir/sharding_param.cc]: I think these comments are a little too verbose. Same for other new comments.

File: xla/python/ifrt/ir/sharding_param.cc [hunk 6]
```diff
@@ -246,10 +271,25 @@ absl::Status ShardingParam::verify() const {
   }
   // TODO(b/491122256): Improve validation logic to check if `dim_shards` can be
   // distributed over the device mesh.
-  int total_dim_shards_size = 1;
+  int64_t total_dim_shards_size = 1;
```
Comment (hyeontaek) [inline xla/python/ifrt/ir/sharding_param.cc]: Same; use `int` and simplify the overflow check below.

File: xla/python/ifrt/ir/sharding_param_test.cc [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+TEST(ShardingParamTest, ValidShardingIsAccepted) {
+  ShardingParam::MinorToMajor minor_to_major;
+  minor_to_major.permutation = {0, 1};
+  minor_to_major.axis_sizes = {2, 4};
+
+  ShardingParam sharding(/*dim_shards=*/{2, 4}, minor_to_major);
+
+  EXPECT_TRUE(sharding.verify().ok());
```
Comment (hyeontaek) [inline xla/python/ifrt/ir/sharding_param_test.cc]: EXPECT_OK() is preferred instead of using EXPECT_TRUE(....ok()) because EXPECT_OK() generates more a helpful test failure message.

File: xla/python/ifrt/ir/sharding_param.cc [hunk 8]
```diff
@@ -166,11 +181,21 @@ mlir::LogicalResult ShardingParam::MinorToMajor::verify(
 void ShardingParam::MinorToMajor::ToDeviceList(
```
Comment (hyeontaek) [inline xla/python/ifrt/ir/sharding_param.cc:176]: I think the check in ToDeviceList is redundant when we already perform the overflow check at deserialization time.

**Comment by hyeontaek** [comment]
> axis_sizes and dim_shards here are attacker/client-controlled proto ints

Just to clarify, we suggest only using IFRT Proxy clients and servers run in a single trusted domain with secure RPC connections where there is a single entity who runs both system components. Thus, a user would only be able to crash their own server and client, which is not different from crashing one of their local process. Thus, there is no particular notion of an attacker in this setup.

The overflow check is still useful generally, just not necessarily from a security hardening perspective.

**Comment by prasanna8585** [comment]
Good to know - thanks for the context on the trust model, that changes things.

Still think the fix itself is worth keeping though: turning the CHECK_LE into a clean InvalidArgumentError gives a malformed or buggy client a proper error back instead of taking down the process, which seems useful on its own regardless of the trust boundary question. Test's already in from earlier. Let me know if that's enough to land, or if you'd want it scoped down further.

**Comment by prasanna8585** [comment]
Hi Team ,

Just following up to check if this PR needs any changes or update from my side.

**Review by hyeontaek — CHANGES_REQUESTED** [review]
Sorry, I did not send out my pending review comments earlier.

**Review by hyeontaek — APPROVED** [review]
Thank you!

File: xla/python/ifrt/ir/sharding_param.cc [hunk 9]
```diff
@@ -170,6 +180,7 @@ void ShardingParam::MinorToMajor::ToDeviceList(
   cum_sizes.reserve(axis_sizes.size());
   for (auto size : axis_sizes) {
     cum_sizes.push_back(cum_size);
+    // Already rejected by verify() at deserialization time.
```
Comment (dimitar-asenov) [inline xla/python/ifrt/ir/sharding_param.cc]: This comment seems out of place. What is rejected?

File: xla/python/ifrt/ir/sharding_param.cc [hunk 10]
```diff
@@ -248,8 +259,20 @@ absl::Status ShardingParam::verify() const {
...
+                       " in [", absl::StrJoin(dim_shards(), "x"), "]"));
+    }
+    if (__builtin_mul_overflow(total_dim_shards_size, dim_shard,
+                               &total_dim_shards_size)) {
+      return absl::InvalidArgumentError(absl::StrCat(
+          "`dim_shards` product overflows: ",
+          absl::StrJoin(dim_shards(), "x")));
+    }
   }
+  // `total_dim_shards_size` is positive here, so no zero-divisor check
```
Comment (dimitar-asenov) [inline xla/python/ifrt/ir/sharding_param.cc]: Delete this comment. In general, we can assume that divisors are not 0 and we don't need to explain this. It's true that sometimes the code might not assure this, but that's a separate issue.

**Review by dimitar-asenov — CHANGES_REQUESTED** [review]

**Review by dimitar-asenov — APPROVED** [review]

**Comment by thcmbs** [comment]
Hi, I am seeing some test failures while integrating this PR:

```
//third_party/tensorflow/compiler/xla/python/ifrt/ir:sharding_param_test
Target ran on September 10, 2026 at 8:46:50 PM GMT+2 for 1 min, 40 sec 
Configured Target exited with: DISTRIBUTOR(NON_ZERO_EXIT) CppCompile returned a non-zero exit code

2 Actions exited with: DISTRIBUTOR(NON_ZERO_EXIT) CppCompile returned a non-zero exit code

DISTRIBUTOR(NON_ZERO_EXIT) CppCompile returned a non-zero exit code
No log available for failed target. Showing root cause for build failure.
[third_party/tensorflow/compiler/xla/python/ifrt/ir/sharding_param_test.cc:42](https://cs.corp.google.com/piper///depot/google3/third_party/tensorflow/compiler/xla/python/ifrt/ir/sharding_param_test.cc?l=42&ws=tap-presubmit-server/700265474&snapshot=2):24: error: no member named 'error' in namespace 'tsl'; did you mean 'util::error'?
   42 |               StatusIs(tsl::error::INVALID_ARGUMENT,
      |                        ^~~~~~~~~~
      |                        util::error
[third_party/googletest/googlemock/include/gmock/gmock-matchers.h:5916](https://cs.corp.google.com/piper///depot/google3/third_party/googletest/googlemock/include/gmock/gmock-matchers.h?l=5916&ws=tap-presubmit-server/700265474&snapshot=2):62: note: expanded from macro 'EXPECT_THAT'
 5916 |       ::testing::internal::MakePredicateFormatterFromMatcher(matcher), value)
      |                                                              ^~~~~~~
[third_party/googletest/googletest/include/gtest/gtest_pred_impl.h:109](https://cs.corp.google.com/piper///depot/google3/third_party/googletest/googletest/include/gtest/gtest_pred_impl.h?l=109&ws=tap-presubmit-server/700265474&snapshot=2):23: note: expanded from macro 'EXPECT_PRED_FORMAT1'
  109 |   GTEST_PRED_FORMAT1_(pred_format, v1, GTEST_NONFATAL_FAILURE_)
      |                       ^~~~~~~~~~~
[third_party/googletest/googletest/include/gtest/gtest_pred_impl.h:100](https://cs.corp.google.com/piper///depot/google3/third_party/googletest/googletest/include/gtest/gtest_pred_impl.h?l=100&ws=tap-presubmit-server/700265474&snapshot=2):17: note: expanded from macro 'GTEST_PRED_FORMAT1_'
  100 |   GTEST_ASSERT_(pred_format(#v1, v1), on_failure)
      |                 ^~~~~~~~~~~
[third_party/googletest/googletest/include/gtest/gtest_pred_impl.h:79](https://cs.corp.google.com/piper///depot/google3/third_party/googletest/googletest/include/gtest/gtest_pred_impl.h?l=79&ws=tap-presubmit-server/700265474&snapshot=2):52: note: expanded from macro 'GTEST_ASSERT_'
   79 |   if (const ::testing::AssertionResult gtest_ar = (expression)) \
      |                                                    ^~~~~~~~~~
[blaze-out/arm-fastbuild/bin/util/task/codes.proto.h:47](https://cs.corp.google.com/piper///depot/google3/blaze-out/arm-fastbuild/bin/util/task/codes.proto.h?l=47&ws=tap-presubmit-server/700265474&snapshot=2):11: note: 'util::error' declared here
   47 | namespace error {
      |           ^
[third_party/tensorflow/compiler/xla/python/ifrt/ir/sharding_param_test.cc:54](https://cs.corp.google.com/piper///depot/google3/third_party/tensorflow/compiler/xla/python/ifrt/ir/sharding_param_test.cc?l=54&ws=tap-presubmit-server/700265474&snapshot=2):24: error: no member named 'error' in namespace 'tsl'; did you mean 'util::error'?
   54 |               StatusIs(tsl::error::INVALID_ARGUMENT,
      |                        ^~~~~~~~~~
      |                        util::error
[third_party/googletest/googlemock/include/gmock/gmock-matchers.h:5916](https://cs.corp.google.com/piper///depot/google3/third_party/googletest/googlemock/include/gmock/gmock-matchers.h?l=5916&ws=tap-presubmit-server/700265474&snapshot=2):62: note: expanded from macro 'EXPECT_THAT'
 5916 |       ::testing::internal::MakePredicateFormatterFromMatcher(matcher), value)
      |                                                              ^~~~~~~
[third_party/googletest/googletest/include/gtest/gtest_pred_impl.h:109](https://cs.corp.google.com/piper///depot/google3/third_party/googletest/googletest/include/gtest/gtest_pred_impl.h?l=109&ws=tap-presubmit-server/700265474&snapshot=2):23: note: expanded from macro 'EXPECT_PRED_FORMAT1'
  109 |   GTEST_PRED_FORMAT1_(pred_format, v1, GTEST_NONFATAL_FAILURE_)
      |                       ^~~~~~~~~~~
[third_party/googletest/googletest/include/gtest/gtest_pred_impl.h:100](https://cs.corp.google.com/piper///depot/google3/third_party/googletest/googletest/include/gtest/gtest_pred_impl.h?l=100&ws=tap-presubmit-server/700265474&snapshot=2):17: note: expanded from macro 'GTEST_PRED_FORMAT1_'
  100 |   GTEST_ASSERT_(pred_format(#v1, v1), on_failure)
      |                 ^~~~~~~~~~~
[third_party/googletest/googletest/include/gtest/gtest_pred_impl.h:79](https://cs.corp.google.com/piper///depot/google3/third_party/googletest/googletest/include/gtest/gtest_pred_impl.h?l=79&ws=tap-presubmit-server/700265474&snapshot=2):52: note: expanded from macro 'GTEST_ASSERT_'
   79 |   if (const ::testing::AssertionResult gtest_ar = (expression)) \
      |                                                    ^~~~~~~~~~
[blaze-out/arm-fastbuild/bin/util/task/codes.proto.h:47](https://cs.corp.google.com/piper///depot/google3/blaze-out/arm-fastbuild/bin/util/task/codes.proto.h?l=47&ws=tap-presubmit-server/700265474&snapshot=2):11: note: 'util::error' declared here
   47 | namespace error {
      |           ^
[third_party/tensorflow/compiler/xla/python/ifrt/ir/sharding_param_test.cc:66](https://cs.corp.google.com/piper///depot/google3/third_party/tensorflow/compiler/xla/python/ifrt/ir/sharding_param_test.cc?l=66&ws=tap-presubmit-server/700265474&snapshot=2):24: error: no member named 'error' in namespace 'tsl'; did you mean 'util::error'?
   66 |               StatusIs(tsl::error::INVALID_ARGUMENT,
      |                        ^~~~~~~~~~
      |                        util::error
[third_party/googletest/googlemock/include/gmock/gmock-matchers.h:5916](https://cs.corp.google.com/piper///depot/google3/third_party/googletest/googlemock/include/gmock/gmock-matchers.h?l=5916&ws=tap-presubmit-server/700265474&snapshot=2):62: note: expanded from macro 'EXPECT_THAT'
 5916 |       ::testing::internal::MakePredicateFormatterFromMatcher(matcher), value)
      |                                                              ^~~~~~~
[third_party/googletest/googletest/include/gtest/gtest_pred_impl.h:109](https://cs.corp.google.com/piper///depot/google3/third_party/googletest/googletest/include/gtest/gtest_pred_impl.h?l=109&ws=tap-presubmit-server/700265474&snapshot=2):23: note: expanded from macro 'EXPECT_PRED_FORMAT1'
  109 |   GTEST_PRED_FORMAT1_(pred_format, v1, GTEST_NONFATAL_FAILURE_)
      |                       ^~~~~~~~~~~
[third_party/googletest/googletest/include/gtest/gtest_pred_impl.h:100](https://cs.corp.google.com/piper///depot/google3/third_party/googletest/googletest/include/gtest/gtest_pred_impl.h?l=100&ws=tap-presubmit-server/700265474&snapshot=2):17: note: expanded from macro 'GTEST_PRED_FORMAT1_'
  100 |   GTEST_ASSERT_(pred_format(#v1, v1), on_failure)
      |                 ^~~~~~~~~~~
[third_party/googletest/googletest/include/gtest/gtest_pred_impl.h:79](https://cs.corp.google.com/piper///depot/google3/third_party/googletest/googletest/include/gtest/gtest_pred_impl.h?l=79&ws=tap-presubmit-server/700265474&snapshot=2):52: note: expanded from macro 'GTEST_ASSERT_'
   79 |   if (const ::testing::AssertionResult gtest_ar = (expression)) \
      |                                                    ^~~~~~~~~~
[blaze-out/arm-fastbuild/bin/util/task/codes.proto.h:47](https://cs.corp.google.com/piper///depot/google3/blaze-out/arm-fastbuild/bin/util/task/codes.proto.h?l=47&ws=tap-presubmit-server/700265474&snapshot=2):11: note: 'util::error' declared here
   47 | namespace error {
      |           ^
3 errors generated.
```

PTAL.

File: xla/python/ifrt/ir/sharding_param_test.cc [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+#include "xla/python/ifrt/ir/sharding_param.h"
+
+#include <limits>
+#include <vector>
+
+#include <gtest/gtest.h>
+#include "absl/status/status_matchers.h"
+#include "llvm/ADT/SmallVector.h"
+
```
Comment (hyeontaek) [inline xla/python/ifrt/ir/sharding_param_test.cc:25]: Please add `#include "xla/tsl/platform/errors.h"` to resolve the integration error.

**Review by hyeontaek — CHANGES_REQUESTED** [review]

File: xla/python/ifrt/ir/sharding_param_test.cc (hunk 11 above)
Comment (prasanna8585) [reply xla/python/ifrt/ir/sharding_param_test.cc:25]: Thanks for catching this pushed a fix. Added the missing xla/tsl/platform/errors.h include and the corresponding //xla/tsl/platform:errors BUILD dependency (the include alone isn't enough for the Bazel target). PTAL.

**Review by hyeontaek — APPROVED** [review]

**Comment by prasanna8585** [comment]
Hi Team,
Are we ready to merge or need any changes from my side?

**Comment by hyeontaek** [comment]
@dimitar-asenov Could you approve this PR again to retry the integration?

**Review by dimitar-asenov — APPROVED** [review]
